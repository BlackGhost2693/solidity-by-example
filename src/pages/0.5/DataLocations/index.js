import React from "react"
import Example from "../../../components/Example"
import html from "./index.html.js"

export default function (props) {
  return (
    <Example
      title="Data Locations - Storage, Memory and Calldata"
      description="Data locations - storage, memory and calldata"
      html={html}
    />
  )
}
