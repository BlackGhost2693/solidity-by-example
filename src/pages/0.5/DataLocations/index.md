Variables are declared as either `storage`, `memory` or `calldata` to explicitly
specify the location of the data.

- `storage` - variable is a state variable (store on blockchain)
- `memory` - variable is in memory and it exists while a function is being called
- `calldata` - special data location that contains function arguments, only available for `external` functions

```solidity
{{{DataLocations}}}
```
