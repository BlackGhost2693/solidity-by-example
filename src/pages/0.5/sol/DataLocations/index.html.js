// metadata
export const version = "0.5.16"
export const title = "Data Locations - Storage, Memory and Calldata"
export const description = "Data locations - storage, memory and calldata"

const html = `<p>Variables are declared as either <code>storage</code>, <code>memory</code> or <code>calldata</code> to explicitly
specify the location of the data.</p>
<ul>
<li><code>storage</code> - variable is a state variable (store on blockchain)</li>
<li><code>memory</code> - variable is in memory and it exists while a function is being called</li>
<li><code>calldata</code> - special data location that contains function arguments, only available for <code>external</code> functions</li>
</ul>
<pre><code class="language-solidity">pragma solidity ^0.5.16;

contract DataLocations {
    uint[] public arr;
    mapping(uint =&gt; address) map;
    struct MyStruct {
        uint foo;
    }
    mapping(uint =&gt; MyStruct) myStructs;

    function f() public {
        // call _f with state variables
        _f(arr, map, myStructs[1]);

        // get a struct from a mapping
        MyStruct storage myStruct = myStructs[1];
        // create a struct in memory
        MyStruct memory myMemStruct = MyStruct(0);
    }

    function _f(
        uint[] storage _arr,
        mapping(uint =&gt; address) storage _map,
        MyStruct storage _myStruct
    ) internal {
        // do something with storage variables
    }

    // You can return memory variables
    function g(uint[] memory _arr) public returns (uint[] memory) {
        // do something with memory array
    }

    function h(uint[] calldata _arr) external {
        // do something with calldata array
    }
}
</code></pre>
`

export default html
