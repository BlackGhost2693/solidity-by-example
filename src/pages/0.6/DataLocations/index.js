import React from "react"
import Example from "../../../components/Example"
import html from "./index.html.js"
import { SOL_VERSION } from "../constants"

export default function (props) {
  return (
    <Example
      title="Data Locations - Storage, Memory and Calldata"
      version={SOL_VERSION}
      description="Data locations - storage, memory and calldata"
      html={html}
    />
  )
}
