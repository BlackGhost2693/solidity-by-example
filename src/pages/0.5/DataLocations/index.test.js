import React from "react"
import { shallow } from "enzyme"
import DataLocations from "./index"

test("it renders", () => {
  const component = shallow(<DataLocations />)

  expect(component).toMatchSnapshot()
})
