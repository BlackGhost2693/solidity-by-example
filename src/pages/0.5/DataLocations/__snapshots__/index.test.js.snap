// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<Example
  description="Data locations - storage, memory and calldata"
  html="<p>Variables are declared as either <code>storage</code>, <code>memory</code> or <code>calldata</code> to explicitly
specify the location of the data.</p>
<ul>
<li><code>storage</code> - variable is a state variable (store on blockchain)</li>
<li><code>memory</code> - variable is in memory and it exists while a function is being called</li>
<li><code>calldata</code> - special data location that contains function arguments, only available for <code>external</code> functions</li>
</ul>
<pre><code class=\\"language-solidity\\">pragma solidity ^0.5.16;

contract DataLocations {
    uint[] public arr;
    mapping(uint =&gt; address) map;
    struct MyStruct {
        uint foo;
    }
    mapping(uint =&gt; MyStruct) myStructs;

    function f() public {
        // call _f with state variables
        _f(arr, map, myStructs[1]);

        // get a struct from a mapping
        MyStruct storage myStruct = myStructs[1];
        // create a struct in memory
        MyStruct memory myMemStruct = MyStruct(0);
    }

    function _f(
        uint[] storage _arr,
        mapping(uint =&gt; address) storage _map,
        MyStruct storage _myStruct
    ) internal {
        // do something with storage variables
    }

    // You can return memory variables
    function g(uint[] memory _arr) public returns (uint[] memory) {
        // do something with memory array
    }

    function h(uint[] calldata _arr) external {
        // do something with calldata array
    }
}
</code></pre>
"
  title="Data Locations - Storage, Memory and Calldata"
/>
`;
